[RFC] Fix coverity issues. (4)

elmart opened the pull request:
Fix 6 coverity issues in "High Impact Outstanding" category.
4 more left for other PR.

elmart committed:
coverity/13762: Out-of-bounds read: RI.

Problem    : Out-of-bounds read @ 2213.
Diagnostic : Real issue.
Rationale  : Error occurs if cmap == ARRAY_SIZE(prt_ps_mbfonts),
             but code takes the `if (prt_out_mbyte)` branch. That's it,
             if a matching encoding is found but not a matching charset.
             In that case, the first matching encoding is used.
Resolution : Remember the value of cmap for the first matching encoding.
             Reset cmap to that value if first matching encoding is
             going to be used.

elmart committed:
coverity/13764: Out-of-bounds read: RI.

Problem    : Out-of-bounds read @ 9514.
Diagnostic : Real issue.
Rationale  : PFD_NOTSPECIAL (253) is defined as the maximum not-special
             value a prefix can have. But stack (and other) arrays are
             defined as having MAXWLEN (250) items.
Resolution : Define MAXWLEN = 254.

elmart committed:
coverity/13773: Resource leak: RI.

Problem    : Resource leak @ 3324.
Diagnostic : Real issue.
Rationale  : Stack is not being freed on error cases.
Resolution : Free stack before invoking EMSG_RET_NULL.

elmart commented:
@tarruda Last commit implements your suggestions on https://github.com/neovim/neovim/issues/1938#issuecomment-72906671. Maybe I have misinterpreted something. I'm having a clang-asan error because of channel_teardown attempting to free an already freed channel. Now, my question is: shouldn't we be removing channel from channels map just before  freeing a channel?

fwalch commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -312,6 +315,9 @@ bool channel_close(uint64_t id)
   }
 
   close_channel(channel);
+  if (!kv_size(channel->call_stack)) {
```
Is it possible to call `maybe_free_channel` here instead?

elmart replied:
Well, yes. In fact I had done it that way the first time. I did that optimization later, as at that point, we know channel is not closed, so we can save checking one condition (and the function call).
But I have no problem in changing it back to previous situation, if its considered better from a readability point of view.

fwalch replied:
Personally I lean towards changing it, but let's see what others think.

elmart replied:
Honestly, me too. I'd prefer just one single place where real freeing is done.
I just thought others would not like checking redundant conditions.
But if nobody opposes, I'll change it so that only one place encapsulates conditions to be able to free, and only one place calls channel_free.

oni-link commented:
@elmart, a channel from a job can make problems when `channel_teardown()` is called, because the callback `job_exit()` will also try to use the (already freed) channel pointer.

elmart commented:
You mean we should just remove the free_channel call within channel_teardown?

elmart commented:
Instead of that, maybe we could use the channel being in the map or not as a flag to denote it thas already been freed. I mean, before freeing, we can check:
1) if it's not closed
2) if call stack size is 0
3) if it's still in the channels map

What I don't understand about the current asan error is: 
If a channel has already been freed, it should have been removed from map. Then, how is it that channel_teardown still finds it within the map, so that it tries to free it again?

oni-link commented:
> What I don't understand about the current asan error is:
If a channel has already been freed, it should have been removed from map. Then, how is it that channel_teardown still finds it within the map, so that it tries to free it again?

Channel from job: If a job terminates, callback `job_exit()` is always called. This keeps the channel in the channel map, but can free the channel.

elmart commented:
I must be misunderstanding something. `job_exit`, to free the channel, calls `free_channel`, which always removes it from the map. Isn't that true?

oni-link commented:
You are right, I only thought of `free(channel)`.

tarruda commented:
Since its difficult to determine when a channel can be freed, maybe we should use reference count to manage its lifetime?

The reference count is increased when:

- channel is created
- entering parse_msgpack
- entering channel_send_call

and decreased when:

- channel is closed
- leaving parse_msgpack
- leaving channel_send_call

The channel is freed when the refcount reaches 0.

elmart commented:
Ok. That seems cleaner and easier to understand. I'll give that a try later.

tarruda commented:
> Ok. That seems cleaner and easier to understand. I'll give that a try later.

Ok. When you do, see if this patch helps:

```diff
diff --git a/src/nvim/msgpack_rpc/channel.c b/src/nvim/msgpack_rpc/channel.c
index 920274f..1782258 100644
--- a/src/nvim/msgpack_rpc/channel.c
+++ b/src/nvim/msgpack_rpc/channel.c
@@ -45,6 +45,7 @@ typedef struct {

 typedef struct {
   uint64_t id;
+  size_t refcount;
   size_t pending_requests;
   PMap(cstr_t) *subscribed_events;
   bool is_job, closed;
@@ -130,6 +131,7 @@ uint64_t channel_from_job(char **argv)
 {
   Channel *channel = register_channel();
   channel->is_job = true;
+  incref(channel);  // job channels are only closed by the exit_cb

   int status;
   channel->data.job = job_start(argv,
@@ -233,6 +235,7 @@ Object channel_send_call(uint64_t id,
     return NIL;
   }

+  incref(channel);
   uint64_t request_id = channel->next_request_id++;
   // Send the msgpack-rpc request
   send_request(channel, request_id, method_name, args);
@@ -251,13 +254,10 @@ Object channel_send_call(uint64_t id,
     return NIL;
   }

-  if (!kv_size(channel->call_stack) && channel->closed) {
-    free_channel(channel);
-  }
-
   if (!channel->pending_requests) {
     send_delayed_notifications();
   }
+  decref(channel);

   return frame.result;
 }
@@ -320,6 +320,7 @@ bool channel_close(uint64_t id)
 static void channel_from_stdio(void)
 {
   Channel *channel = register_channel();
+  incref(channel);  // stdio channels are only closed on exit
   channel->is_job = false;
   // read stream
   channel->data.streams.read = rstream_new(parse_msgpack,
@@ -354,23 +355,18 @@ static void job_err(RStream *rstream, void *data, bool eof)

 static void job_exit(Job *job, void *data)
 {
-  Channel *channel = data;
-  // ensure the channel is flagged as closed so channel_send_call frees it
-  // later
-  channel->closed = true;
-  if (!kv_size(channel->call_stack)) {
-    free_channel(channel);
-  }
+  decref(data);
 }

 static void parse_msgpack(RStream *rstream, void *data, bool eof)
 {
   Channel *channel = data;
+  incref(channel);

   if (eof) {
     close_channel(channel);
     call_set_error(channel, "Channel was closed by the client");
-    return;
+    goto end;
   }

   size_t count = rstream_pending(rstream);
@@ -408,7 +404,7 @@ static void parse_msgpack(RStream *rstream, void *data, bool eof)
       }
       msgpack_unpacked_destroy(&unpacked);
       // Bail out from this event loop iteration
-      return;
+      goto end;
     }

     handle_request(channel, &unpacked.data);
@@ -431,6 +427,9 @@ static void parse_msgpack(RStream *rstream, void *data, bool eof)
                            "This error can also happen when deserializing "
                            "an object with high level of nesting");
   }
+
+end:
+  decref(channel);
 }

 static void handle_request(Channel *channel, msgpack_object *request)
@@ -450,7 +449,7 @@ static void handle_request(Channel *channel, msgpack_object *request)
                                          &out_buffer))) {
       char buf[256];
       snprintf(buf, sizeof(buf),
-               "Channel %" PRIu64 " sent an invalid message, closing.",
+               "Channel %" PRIu64 " sent an invalid message, closed.",
                channel->id);
       call_set_error(channel, buf);
     }
@@ -628,6 +627,7 @@ static void close_channel(Channel *channel)
   }

   channel->closed = true;
+
   if (channel->is_job) {
     if (channel->data.job) {
       job_stop(channel->data.job);
@@ -638,15 +638,17 @@ static void close_channel(Channel *channel)
     uv_handle_t *handle = (uv_handle_t *)channel->data.streams.uv;
     if (handle) {
       uv_close(handle, close_cb);
-      free_channel(channel);
     } else {
-      event_push((Event) { .handler = on_stdio_close }, false);
+      event_push((Event) { .handler = on_stdio_close, .data = channel }, false);
     }
   }
+
+  decref(channel);
 }

 static void on_stdio_close(Event e)
 {
+  decref(e.data);
   mch_exit(0);
 }

@@ -679,6 +681,7 @@ static void close_cb(uv_handle_t *handle)
 static Channel *register_channel(void)
 {
   Channel *rv = xmalloc(sizeof(Channel));
+  rv->refcount = 1;
   rv->closed = false;
   rv->unpacker = msgpack_unpacker_new(MSGPACK_UNPACKER_INIT_BUFFER_SIZE);
   rv->id = next_id++;
@@ -787,6 +790,18 @@ static void send_delayed_notifications(void)
   }
 }

+static void incref(Channel *channel)
+{
+  channel->refcount++;
+}
+
+static void decref(Channel *channel)
+{
+  if (!(--channel->refcount)) {
+    free_channel(channel);
+  }
+}
+
 #if MIN_LOG_LEVEL <= DEBUG_LOG_LEVEL
 #define REQ "[request]      "
 #define RES "[response]     "
```

tarruda commented:
With the above patch, closing a channel will:

- decref  to match the `refcount = 1` set on creation
- set a flag that blocks the channel from being used in the future.
- force a connection close or job

Also, a channel is closed when:

  - `channel_close` is called
  - the connection is closed by the other end(EOF)
  - an error condition is encountered

Since refcount can only be 0 if the channel is closed, we no longer need `channel->closed` checks to determine if `free_channel` should be called.

chrisbra commented:
On Fr, 06 Feb 2015, Eliseo Martínez wrote:

> Fix 5 coverity issues in "High Impact Outstanding" category.
> 5 more left for other PR.
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/neovim/neovim/pull/1947
> 
> -- Commit Summary --
> 
>   * coverity/13762: Out-of-bounds read: FP.
>   * coverity/13764: Out-of-bounds read: RI.
>   * coverity/13773: Resource leak: RI.
>   * coverity/{68484,68485}: Read from pointer after free: RI.
> 
> -- File Changes --
> 
>     M src/nvim/hardcopy.c (7)
>     M src/nvim/msgpack_rpc/channel.c (25)
>     M src/nvim/regexp_nfa.c (16)
>     M src/nvim/spell.c (2)
> 
> -- Patch Links --
> 
> https://github.com/neovim/neovim/pull/1947.patch
> https://github.com/neovim/neovim/pull/1947.diff

Would you mind, reporting those issues that affect vim back to vim-dev?
Thanks.

Best,
Christian
-- 
Wer den Lebensbecher bis auf den Grund ausleeren will, muß sich 
vernünftigerweise auf die gewöhnliche Hefe gefaßt machen.
		-- Karl Julius Weber

elmart commented:
@chrisbra 
> Would you mind, reporting those issues that affect vim back to vim-dev?

I'm always glad to contribute back to vim. I've done it in the past and I will continue doing it for particularly important issues. Note, however, that I probably can't do it for every small issue we fix here. For, as codebases diverge, it becomes more and more difficult to know if some issue affects vim too, and to port corresponding changes. Particularly for coverity issues, as they tend to be obscure corner cases not easily reproducible in the running thing. The fact that vim's coverity build hasn't been updated for a long time doesn't help, either.

So, in all honesty, I think the optimal thing for best mutual benefit would be for someone at vim-dev, with proper knowledge of what applies there and what doesn't, to monitor our commits and port relevant changes there. This is, the same we do with vim-patches, but the other way. For coverity issues, we label all commits fixing them with the prefix "coverity/nnnnn: ", and the ones that are diagnosed as real issues have the suffix "RI". For other kind of bugs, they all have the GitHub label "bug". That would make reasonably simple for someone at vim-dev to follow at least those.

That said, I insist: For particularly important issues, I will continue sending patches to vim-dev or, at least, reporting them.

tarruda committed:
coverity/{68484,68485}: Read from pointer after free: RI.

Problem    : Read from pointer after free @ {242, 391}.
Diagnostic : Real issues.
Rationale  : Channel gets indeed freed on error case, producing
             incorrect accesses to freed pointer later on.
Resolution : Implement reference counting mechanism to know when to free
             channel.

elmart commented:
@tarruda I applied your patch, and it seems to work (at least, coverity doesn't complain anymore). I don't fully understand, though, why the return within `if (frame.errored)` at `channel_send_call` doesn't need to decref.

Also, travis build is failing because a problem with functional tests on OSX & gcc:
```
PANIC: unprotected error in call to Lua API (29)
CMake Error at /Users/travis/build/neovim/neovim/cmake/RunTests.cmake:25 (message):
  Running functional tests failed with error: 1.
```

elmart commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -251,13 +259,10 @@ Object channel_send_call(uint64_t id,
     return NIL;
```
I think we need a `decref` before this `return`. Does anyone disagree?

oni-link replied:
I would have expected a `decref()` in the `if (frame.errored) {...}` block but not here. To which `incref()` would this `decref()` correspond?

elmart replied:
Oh, sorry. I annotated the wrong return. I meant the one withing `if (frame.errored)`, as I asked at the beginning in the main conversation thread.

elmart replied:
In fact, there's something weird going on here (Github bug). I think I annotated the correct `return` (the one within `if (frame.errored)`. And so I see in the diff comment at main conversation page. But, for some reason, it is placed in the other return when going into the code. Strange.

In any case, the important thing is we do agree there's a missing decref within `if (frame.errored)`.

elmart commented:
Added review commit to address @oni-link's comments.

tarruda commented:
> @tarruda I applied your patch, and it seems to work (at least, coverity doesn't complain anymore). I don't fully understand, though, why the return within if (frame.errored) at channel_send_call doesn't need to decref.

:+1:  I forgot about it

elmart committed:
coverity/102149: Out-of-bounds access: FP.

Problem    : Out-of-bounds access @ 5815.
Diagnostic : False positive.
Rationale  : Error occurs when event_name2nr() returns NUM_EVENTS, which
             means an event with that name was not found. That cannot
             happen, as previous check using find_end_event() @ 5744
             ensures event name exists.
Resolution : Assert event_name2nr() result is less thatn NUM_EVENTS.

elmart commented:
Added review commit.

elmart commented:
Review commits integrated.
Added newly fixed issue (last commit).

tarruda commented:
Something is causing the OSX builds to run very slow, crossing the 50 minutes limit

elmart commented:
Yes. I've restarted this PR's builds a couple of times, but to no result.
Do you comment that in this PR for any reason?

tarruda commented:
> Do you comment that in this PR for any reason?

No its not because of this PR, I've experienced this problem all day. I think its a travis issue(like the OSX vms suddenly slowed down)

tarruda commented:
The msgpack-rpc bits LGTM.

oni-link commented:
LGTM.

elmart commented:
Ok, guys.
@tarruda Thanks very much for your help.
@oni-link Thansk very much for your help, and as always, thanks very much for your detailed review. Please consider opening a PR to complete whatever work you estimate pending on https://github.com/elmart/neovim/commit/6bf03f47d1c8c8e23a61271c67c66d404e6ddd17. I don't feel entitled to do it myself, as I don't grok the involved mechanisms.
@justinmk I'll be merging this after finishing some other thing I'm doing now.

justinmk commented:
:+1:

tarruda commented:
@elmart thanks for your dedication in solving these issues.

As I imagined, something is wrong with travis OSX vms: http://www.traviscistatus.com/incidents/0sc9kyjvbhls

elmart commented:
> As I imagined, something is wrong with travis OSX vms

Aha! Well, that's alleviating. I was starting to fear having to prescind from OSX builds.
Ok, then I think we should subscribe to updates on this issue's status, and make OSX builds optional in the meantime. Or, if you think this is a matter of hours, well, then, leave it as it is. Not sure about that. What's yours on that?

tarruda commented:
> Ok, then I think we should subscribe to updates on this issue's status

I already follow [@traviscistatus](https://twitter.com/traviscistatus), but it seems they took more than 24h to be notified about the issue(I should have notified them as soon as I noticed)
